pkg/node/manager/manager.go

```diff
@@ -226,11 +241,105 @@ func New(c Configuration, ipCache IPCache, nodeMetrics *nodeMetrics, healthRepor
 	return m, nil
 }
 
-func (m *manager) Start(hive.HookContext) error {
+func (m *manager) Start(ctx hive.HookContext) error {
+	if err := m.restoreExistingIdentities(ctx); err != nil {
+		log.WithError(err).Warn("failed to restore existing node CIDR identities")
+	}
 	m.workerpool = workerpool.New(numBackgroundWorkers)
 	return m.workerpool.Submit("backgroundSync", m.backgroundSync)
 }
 
+// restoreExistingIdentities takes any already-existing identities in the BPF ipcache and
+// back-fills them in to the userspace ipcache.
+// This is necessary to prevent policy "blips" on startup by preventing accidental re-use of
+// an identity still in the bpf ipcache and policymaps. It roughly matches the existing
+// CIDR identity restoration flow.
+func (m *manager) restoreExistingIdentities(ctx context.Context) error {
+
+	// Do not restore node identities if they are not selectable by CIDR policy
+	if !option.Config.PolicyCIDRSelectsNodes {
+		return nil
+	}
+
+	// Dump existing ip-identity pairs from the ipcache
+	existingPrefixes := map[netip.Prefix]identity.NumericIdentity{}
+	err := ipcachemap.IPCacheMap().DumpWithCallback(func(key bpf.MapKey, value bpf.MapValue) {
+		k := key.(*ipcachemap.Key)
+		v := value.(*ipcachemap.RemoteEndpointInfo)
+		nid := identity.NumericIdentity(v.SecurityIdentity)
+		if nid.Scope() == identity.IdentityScopeRemoteNode {
+			existingPrefixes[k.Prefix()] = nid
+		}
+	})
+	// DumpWithCallback unconditionally opens the map, we must close it
+	ipcachemap.IPCacheMap().Close()
+	if err != nil {
+		if os.IsNotExist(err) {
+			return nil
+		}
+		return fmt.Errorf("failed to dump ipcache while restoring node CIDR identities: %w", err)
+	}
+
+	if len(existingPrefixes) == 0 {
+		log.Infof("Retrieved no existing remote-node identities from bpf ipcache.")
+		return nil
+	}
+
+	log.WithField("count", len(existingPrefixes)).Infof("Retrieved existing remote-node identities from bpf ipcache, inserting in to userspace")
+
+	m.restoredNodeIdentities = make(map[netip.Prefix]*identity.Identity, len(existingPrefixes))
+	// Insert restored prefixes in to the userspace ipcache and then try to reserve its existing numerical ID
+	for prefix, nid := range existingPrefixes {
+		log := log.WithFields(logrus.Fields{
+			logfields.CIDR:        prefix,
+			logfields.OldIdentity: nid,
+		})
+
+		// Insert a stub metadata entry flagging this as a remote-node.
+		// The ipcache will not process the prefix (and thus will not allocate an identity) because ipcache updates
+		// are blocked until daemon startup is complete.
+		m.ipcache.UpsertMetadata(prefix, source.Restored, restoredResource, labels.LabelRemoteNode)
+
+		// Now that the ipcache knows about the prefix, try to allocate the existing ID number for it.
+		id, err := m.ipcache.AllocateIdentity(ctx, prefix, nid)
```

Comment:
Another concern here if the argument above doesn't sell you, is there a race condition here? Can we guarantee that this start function completes before the ipcache starts?